api/callbacks.go

```diff
@@ -247,27 +248,26 @@ func cScan(ptr *C.db_t, gasMeter *C.gas_meter_t, usedGas *C.uint64_t, start C.Bu
 	gasAfter := gm.GasConsumed()
 	*usedGas = (C.uint64_t)((gasAfter - gasBefore) * GasMultiplier)
 
-	// Let's hope this works!
 	*out = buildIterator(iter, gasMeter)
 	return C.GoResult_Ok
 }
 
 //export cNext
-func cNext(ptr *C.iterator_t, gasMeter *C.gas_meter_t, usedGas *C.uint64_t, key *C.Buffer, val *C.Buffer) (ret C.GoResult) {
+func cNext(idx C.iterator_t, gasMeter *C.gas_meter_t, usedGas *C.uint64_t, key *C.Buffer, val *C.Buffer) (ret C.GoResult) {
 	// typical usage of iterator
 	// 	for ; itr.Valid(); itr.Next() {
 	// 		k, v := itr.Key(); itr.Value()
 	// 		...
 	// 	}
 
 	defer recoverPanic(&ret)
-	if ptr == nil || gasMeter == nil || usedGas == nil || key == nil || val == nil {
+	if idx == 0 || gasMeter == nil || usedGas == nil || key == nil || val == nil {
 		// we received an invalid pointer
 		return C.GoResult_BadArgument
 	}
 
 	gm := *(*GasMeter)(unsafe.Pointer(gasMeter))
-	iter := *(*dbm.Iterator)(unsafe.Pointer(ptr))
+	iter := retrieveIterator(int(idx))
```

Comment:
Are Go integers guaranteed to be 64 bit? I think that isn't the case, but there is a guaranteed-64-bit type iirc. you should use that instead (in the signatures of `retrieveIterator` and `storeIterator` as well

Reply:
The problem is `len(stack)` returns `int`. Not `usize`, not `int64`... We can cast it there, but it is an int.

Also, int is defined to be the default width of the architecture - 32 or 64 bits. We only support 64 bit architectures with wasmd, so this is kind of safe. But I agree, sub-optimial.

If I use the two counter approach, I can make this explicit u32

Reply:
if `len` returns `int`, then we should convert to a known-width integer to ensure correct memory representation. i think `uint32` is more than enough, no way anyone will have enough gas to do more than that.

Reply:
I can see you went with 64-bit integers. cool :+1: